File: go/solutions/day10.go
package solutions

import (
	"fmt"
	"strconv"
	"strings"

	"main.go/go/helpers"
)

type crt struct {
	cycle                int
	register             int
	image                [][]string
	sumOfSignalStrengths int
}

func (c *crt) performCycle() {
	if c.cycle%40 == 20 {
		c.sumOfSignalStrengths += c.signalStrength()
	}
	c.cycle++
	c.draw()
}

func (c *crt) draw() {
	pixel := "."
	if c.spriteVisible() {
		pixel = "#"
	}
	if len(c.image) == 0 || len(*c.lastRow()) == 40 {
		c.image = append(c.image, []string{pixel})
	} else {
		*c.lastRow() = append(*c.lastRow(), pixel)
	}
}

func (c crt) lastRow() *[]string {
	return &c.image[len(c.image)-1]
}

func (c crt) spritePosition() map[int]bool {
	middle := c.register % 40
	return map[int]bool{
		middle - 1: true, middle: true, middle + 1: true,
	}
}

func (c crt) spriteVisible() bool {
	sp := c.spritePosition()
	if len(c.image) == 0 {
		return sp[0]
	}
	lr := len(*c.lastRow())
	return sp[lr] || (sp[0] && lr == 40)
}

func (c crt) signalStrength() int {
	return c.cycle * c.register
}

func (c crt) print() {
	for _, row := range c.image {
		fmt.Println(strings.Join(row, ""))
	}
}

func (c crt) toString() string {
	rows := []string{}
	for _, row := range c.image {
		rows = append(rows, strings.Join(row, ""))
	}
	return strings.Join(rows, "\n")
}

func parseInstruction(i string) (string, int) {
	parts := strings.Split(i, " ")
	operation := parts[0]
	if operation == "noop" {
		return operation, 0
	}
	value, _ := strconv.Atoi(parts[1])
	return operation, value
}

func Day10() (string, string) {
	file, scanner := helpers.GetFile(10)
	defer file.Close()

	c := crt{cycle: 1, register: 1}
	for scanner.Scan() {
		c.performCycle()

		operation, value := parseInstruction(scanner.Text())
		if operation == "addx" {
			c.performCycle()
			c.register += value
		}
	}
	return strconv.Itoa(c.sumOfSignalStrengths), c.toString()
}

var Day10part2Answer = `####.#..#.###..###..####.####..##..#....
...#.#..#.#..#.#..#.#....#....#..#.#....
..#..#..#.#..#.#..#.###..###..#....#....
.#...#..#.###..###..#....#....#....#....
#....#..#.#....#.#..#....#....#..#.#....
####..##..#....#..#.#....####..##..####.`

File: python/day10.py
from dataclasses import dataclass, field

from utils import get_data


class CRT:
    cycle = 1
    row_length = 40
    register = 1
    sum_of_significant_signals = 0

    def __init__(self):
        self.rows = [[]]

    def parse_instruction(self, instruction):
        parts = instruction.split(' ')
        if parts[0] == 'noop':
            return parts[0], 0
        return parts[0], int(parts[-1])

    def perform_cycle(self):
        if self.cycle % 40 == 20:
            self.sum_of_significant_signals += self.signal_strength
        self.draw()
        self.cycle += 1

    def draw(self):
        pixel = '#' if self.sprite_visible else '.'
        if len(self.rows[-1]) == self.row_length:
            self.rows.append([pixel])
        else:
            self.rows[-1].append(pixel)

    @property
    def signal_strength(self):
        return self.cycle * self.register

    @property
    def sprite_position(self):
        middle = self.register % 40
        return (middle - 1, middle, middle + 1)

    @property
    def sprite_visible(self):
        return len(self.rows[-1]) in self.sprite_position

    def print(self):
        for row in self.rows:
            print(''.join(row))


def main():
    crt = CRT()
    instructions = get_data(10)
    for instruction in instructions:
        crt.perform_cycle()

        operation, value = crt.parse_instruction(instruction)
        if operation == 'addx':
            crt.perform_cycle()
            crt.register += value

    crt.print()
    return crt.sum_of_significant_signals, crt.rows


if __name__ == '__main__':
    main()


part2_answer = [
    ['#', '#', '#', '#', '.', '#', '.', '.', '#', '.', '#', '#', '#', '.', '.', '#', '#', '#', '.', '.',
        '#', '#', '#', '#', '.', '#', '#', '#', '#', '.', '.', '#', '#', '.', '.', '#', '.', '.', '.', '.'],
    ['.', '.', '.', '#', '.', '#', '.', '.', '#', '.', '#', '.', '.', '#', '.', '#', '.', '.', '#', '.',
        '#', '.', '.', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '#', '.', '#', '.', '.', '.', '.'],
    ['.', '.', '#', '.', '.', '#', '.', '.', '#', '.', '#', '.', '.', '#', '.', '#', '.', '.', '#', '.',
        '#', '#', '#', '.', '.', '#', '#', '#', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '.', '.'],
    [
        '.', '#', '.', '.', '.', '#', '.', '.', '#', '.', '#', '#', '#', '.', '.', '#', '#', '#', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '.', '.'],
    ['#', '.', '.', '.', '.', '#', '.', '.', '#', '.', '#', '.', '.', '.', '.', '#', '.', '#', '.', '.',
        '#', '.', '.', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '#', '.', '#', '.', '.', '.', '.'],
    ['.', '#', '#', '#', '.', '.', '#', '#', '.', '.', '#', '.', '.', '.', '.', '#', '.', '.', '#', '.',
        '#', '.', '.', '.', '.', '#', '#', '#', '#', '.', '.', '#', '#', '.', '.', '#', '#', '#', '#', '.']
]
